File: sample/scala/02_raytrace_spheres.scala
import scala.collection.mutable
import scala.util.boundary, boundary.break
import scala.math.*
import java.nio.file.{Files, Paths}


// 02: Sample that runs a mini sphere-only ray tracer and outputs a PNG image.
// Dependencies are kept minimal (time only) for transpilation compatibility.

def clamp01(v: Double): Double = {
    if (v < 0.0) {
        return 0.0
    }
    if (v > 1.0) {
        return 1.0
    }
    return v
}

def hit_sphere(ox: Double, oy: Double, oz: Double, dx: Double, dy: Double, dz: Double, cx: Double, cy: Double, cz: Double, r: Double): Double = {
    var lx: Double = ox - cx
    var ly: Double = oy - cy
    var lz: Double = oz - cz
    var a: Double = ((dx * dx + dy * dy) + dz * dz)
    var b: Double = (2.0 * (((lx * dx + ly * dy) + lz * dz)))
    var c: Double = (((lx * lx + ly * ly) + lz * lz) - r * r)
    var d: Double = (b * b - (4.0 * a * c))
    if (d < 0.0) {
        return __pytra_float(-1.0)
    }
    var sd: Double = __pytra_float(scala.math.sqrt(__pytra_float(d)))
    var t0: Double = ((((-b) - sd)) / (2.0 * a))
    var t1: Double = ((((-b) + sd)) / (2.0 * a))
    if (t0 > 0.001) {
        return t0
    }
    if (t1 > 0.001) {
        return t1
    }
    return __pytra_float(-1.0)
}

def render(width: Long, height: Long, aa: Long): mutable.ArrayBuffer[Long] = {
    var pixels: mutable.ArrayBuffer[Long] = mutable.ArrayBuffer[Long]()
    var ox: Double = 0.0
    var oy: Double = 0.0
    var oz: Double = __pytra_float(-3.0)
    var lx: Double = __pytra_float(-0.4)
    var ly: Double = 0.8
    var lz: Double = __pytra_float(-0.45)
    var __hoisted_cast_1: Double = __pytra_float(aa)
    var __hoisted_cast_2: Double = __pytra_float(height - 1L)
    var __hoisted_cast_3: Double = __pytra_float(width - 1L)
    var __hoisted_cast_4: Double = __pytra_float(height)
    var y: Long = 0L
    while (y < height) {
        var x: Long = 0L
        while (x < width) {
            var ar: Long = 0L
            var ag: Long = 0L
            var ab: Long = 0L
            var ay: Long = 0L
            while (ay < aa) {
                var ax: Long = 0L
                while (ax < aa) {
                    var fy: Double = (((__pytra_float(y) + ((__pytra_float(ay) + 0.5) / __hoisted_cast_1))) / __hoisted_cast_2)
                    var fx: Double = (((__pytra_float(x) + ((__pytra_float(ax) + 0.5) / __hoisted_cast_1))) / __hoisted_cast_3)
                    var sy: Double = (1.0 - 2.0 * fy)
                    var sx: Double = (((2.0 * fx - 1.0)) * (__pytra_float(width) / __hoisted_cast_4))
                    var dx: Double = sx
                    var dy: Double = sy
                    var dz: Double = 1.0
                    var inv_len: Double = __pytra_float(1.0 / __pytra_float(scala.math.sqrt(__pytra_float((dx * dx + dy * dy) + dz * dz))))
                    dx *= inv_len
                    dy *= inv_len
                    dz *= inv_len
                    var t_min: Double = 1e+30
                    var hit_id: Long = __pytra_int(-1L)
                    var t: Double = hit_sphere(ox, oy, oz, dx, dy, dz, (-0.8), (-0.2), 2.2, 0.8)
                    if ((t > 0.0) && (t < t_min)) {
                        t_min = t
                        hit_id = 0L
                    }
                    t = hit_sphere(ox, oy, oz, dx, dy, dz, 0.9, 0.1, 2.9, 0.95)
                    if ((t > 0.0) && (t < t_min)) {
                        t_min = t
                        hit_id = 1L
                    }
                    t = hit_sphere(ox, oy, oz, dx, dy, dz, 0.0, (-1001.0), 3.0, 1000.0)
                    if ((t > 0.0) && (t < t_min)) {
                        t_min = t
                        hit_id = 2L
                    }
                    var r: Long = 0L
                    var g: Long = 0L
                    var b: Long = 0L
                    if (hit_id >= 0L) {
                        var px: Double = (ox + dx * t_min)
                        var py: Double = (oy + dy * t_min)
                        var pz: Double = (oz + dz * t_min)
                        var nx: Double = 0.0
                        var ny: Double = 0.0
                        var nz: Double = 0.0
                        if (hit_id == 0L) {
                            nx = ((px + 0.8) / 0.8)
                            ny = ((py + 0.2) / 0.8)
                            nz = ((pz - 2.2) / 0.8)
                        } else {
                            if (hit_id == 1L) {
                                nx = ((px - 0.9) / 0.95)
                                ny = ((py - 0.1) / 0.95)
                                nz = ((pz - 2.9) / 0.95)
                            } else {
                                nx = 0.0
                                ny = 1.0
                                nz = 0.0
                            }
                        }
                        var diff: Double = (((nx * (-lx)) + (ny * (-ly))) + (nz * (-lz)))
                        diff = clamp01(diff)
                        var base_r: Double = 0.0
                        var base_g: Double = 0.0
                        var base_b: Double = 0.0
                        if (hit_id == 0L) {
                            base_r = 0.95
                            base_g = 0.35
                            base_b = 0.25
                        } else {
                            if (hit_id == 1L) {
                                base_r = 0.25
                                base_g = 0.55
                                base_b = 0.95
                            } else {
                                var checker: Long = __pytra_int((px + 50.0) * 0.8) + __pytra_int((pz + 50.0) * 0.8)
                                if (checker % 2L == 0L) {
                                    base_r = 0.85
                                    base_g = 0.85
                                    base_b = 0.85
                                } else {
                                    base_r = 0.2
                                    base_g = 0.2
                                    base_b = 0.2
                                }
                            }
                        }
                        var shade: Double = (0.12 + 0.88 * diff)
                        r = __pytra_int(255.0 * clamp01(base_r * shade))
                        g = __pytra_int(255.0 * clamp01(base_g * shade))
                        b = __pytra_int(255.0 * clamp01(base_b * shade))
                    } else {
                        var tsky: Double = (0.5 * (dy + 1.0))
                        r = __pytra_int(255.0 * ((0.65 + 0.2 * tsky)))
                        g = __pytra_int(255.0 * ((0.75 + 0.18 * tsky)))
                        b = __pytra_int(255.0 * ((0.9 + 0.08 * tsky)))
                    }
                    ar += r
                    ag += g
                    ab += b
                    ax += 1L
                }
                ay += 1L
            }
            var samples: Long = aa * aa
            pixels.append(__pytra_int(ar / samples))
            pixels.append(__pytra_int(ag / samples))
            pixels.append(__pytra_int(ab / samples))
            x += 1L
        }
        y += 1L
    }
    return pixels
}

def run_raytrace(): Unit = {
    var width: Long = 1600L
    var height: Long = 900L
    var aa: Long = 2L
    var out_path: String = "sample/out/02_raytrace_spheres.png"
    var start: Double = __pytra_perf_counter()
    var pixels: mutable.ArrayBuffer[Long] = render(width, height, aa)
    __pytra_write_rgb_png(out_path, width, height, pixels)
    var elapsed: Double = __pytra_perf_counter() - start
    __pytra_print("output:", out_path)
    __pytra_print("size:", width, "x", height)
    __pytra_print("elapsed_sec:", elapsed)
}

def main(args: Array[String]): Unit = {
    run_raytrace()
}

File: sample/zig/02_raytrace_spheres.zig
const std = @import("std");
const pytra = @import("built_in/py_runtime.zig");
const math = @import("std/math.zig");
const png = @import("utils/png.zig");
const time = @import("std/time.zig");
const perf_counter = time.perf_counter;

// 02: Sample that runs a mini sphere-only ray tracer and outputs a PNG image.
// Dependencies are kept minimal (time only) for transpilation compatibility.

fn clamp01(v: f64) f64 {
    if ((v < 0.0)) {
        return 0.0;
    }
    if ((v > 1.0)) {
        return 1.0;
    }
    return v;
}

fn hit_sphere(ox: f64, oy: f64, oz: f64, dx: f64, dy: f64, dz: f64, cx: f64, cy: f64, cz: f64, r: f64) f64 {
    const lx: f64 = (ox - cx);
    const ly: f64 = (oy - cy);
    const lz: f64 = (oz - cz);
    
    const a: f64 = (((dx * dx) + (dy * dy)) + (dz * dz));
    const b: f64 = (2.0 * (((lx * dx) + (ly * dy)) + (lz * dz)));
    const c: f64 = ((((lx * lx) + (ly * ly)) + (lz * lz)) - (r * r));
    
    const d: f64 = ((b * b) - ((4.0 * a) * c));
    if ((d < 0.0)) {
        return -1.0;
    }
    const sd: f64 = math.sqrt(d);
    const t0: f64 = ((-b - sd) / (2.0 * a));
    const t1: f64 = ((-b + sd) / (2.0 * a));
    
    if ((t0 > 0.001)) {
        return t0;
    }
    if ((t1 > 0.001)) {
        return t1;
    }
    return -1.0;
}

fn render(width: i64, height: i64, aa: i64) pytra.Obj {
    const pixels: pytra.Obj = pytra.bytearray(0);
    
    // Camera origin
    const ox: f64 = 0.0;
    const oy: f64 = 0.0;
    const oz: f64 = -3.0;
    
    // Light direction (normalized)
    const lx: f64 = -0.4;
    const ly: f64 = 0.8;
    const lz: f64 = -0.45;
    
    var y: i64 = 0;
    while (y < height) : (y += 1) {
        var x: i64 = 0;
        while (x < width) : (x += 1) {
            var ar: i64 = 0;
            var ag: i64 = 0;
            var ab: i64 = 0;
            
            var ay: i64 = 0;
            while (ay < aa) : (ay += 1) {
                var ax: i64 = 0;
                while (ax < aa) : (ax += 1) {
                    const fy: f64 = ((@as(f64, @floatFromInt(y)) + ((@as(f64, @floatFromInt(ay)) + 0.5) / @as(f64, @floatFromInt(aa)))) / @as(f64, @floatFromInt((height - 1))));
                    const fx: f64 = ((@as(f64, @floatFromInt(x)) + ((@as(f64, @floatFromInt(ax)) + 0.5) / @as(f64, @floatFromInt(aa)))) / @as(f64, @floatFromInt((width - 1))));
                    const sy: f64 = (1.0 - (2.0 * fy));
                    const sx: f64 = (((2.0 * fx) - 1.0) * (@as(f64, @floatFromInt(width)) / @as(f64, @floatFromInt(height))));
                    
                    var dx: f64 = sx;
                    var dy: f64 = sy;
                    var dz: f64 = 1.0;
                    const inv_len: f64 = (1.0 / math.sqrt((((dx * dx) + (dy * dy)) + (dz * dz))));
                    dx *= inv_len;
                    dy *= inv_len;
                    dz *= inv_len;
                    
                    var t_min: f64 = 1e+30;
                    var hit_id: i64 = -1;
                    
                    var t: f64 = hit_sphere(ox, oy, oz, dx, dy, dz, -0.8, -0.2, 2.2, 0.8);
                    if (((t > 0.0) and (t < t_min))) {
                        t_min = t;
                        hit_id = 0;
                    }
                    t = hit_sphere(ox, oy, oz, dx, dy, dz, 0.9, 0.1, 2.9, 0.95);
                    if (((t > 0.0) and (t < t_min))) {
                        t_min = t;
                        hit_id = 1;
                    }
                    t = hit_sphere(ox, oy, oz, dx, dy, dz, 0.0, -1001.0, 3.0, 1000.0);
                    if (((t > 0.0) and (t < t_min))) {
                        t_min = t;
                        hit_id = 2;
                    }
                    var r: i64 = 0;
                    var g: i64 = 0;
                    var b: i64 = 0;
                    
                    if ((hit_id >= 0)) {
                        const px: f64 = (ox + (dx * t_min));
                        const py: f64 = (oy + (dy * t_min));
                        const pz: f64 = (oz + (dz * t_min));
                        
                        var nx: f64 = 0.0;
                        var ny: f64 = 0.0;
                        var nz: f64 = 0.0;
                        
                        if ((hit_id == 0)) {
                            nx = ((px + 0.8) / 0.8);
                            ny = ((py + 0.2) / 0.8);
                            nz = ((pz - 2.2) / 0.8);
                        } else {
                            if ((hit_id == 1)) {
                                nx = ((px - 0.9) / 0.95);
                                ny = ((py - 0.1) / 0.95);
                                nz = ((pz - 2.9) / 0.95);
                            } else {
                                nx = 0.0;
                                ny = 1.0;
                                nz = 0.0;
                            }
                        }
                        var diff: f64 = (((nx * -lx) + (ny * -ly)) + (nz * -lz));
                        diff = clamp01(diff);
                        
                        var base_r: f64 = 0.0;
                        var base_g: f64 = 0.0;
                        var base_b: f64 = 0.0;
                        
                        if ((hit_id == 0)) {
                            base_r = 0.95;
                            base_g = 0.35;
                            base_b = 0.25;
                        } else {
                            if ((hit_id == 1)) {
                                base_r = 0.25;
                                base_g = 0.55;
                                base_b = 0.95;
                            } else {
                                const checker: i64 = (@as(i64, @intFromFloat(((px + 50.0) * 0.8))) + @as(i64, @intFromFloat(((pz + 50.0) * 0.8))));
                                if ((@mod(checker, 2) == 0)) {
                                    base_r = 0.85;
                                    base_g = 0.85;
                                    base_b = 0.85;
                                } else {
                                    base_r = 0.2;
                                    base_g = 0.2;
                                    base_b = 0.2;
                                }
                            }
                        }
                        const shade: f64 = (0.12 + (0.88 * diff));
                        r = @as(i64, @intFromFloat((255.0 * clamp01((base_r * shade)))));
                        g = @as(i64, @intFromFloat((255.0 * clamp01((base_g * shade)))));
                        b = @as(i64, @intFromFloat((255.0 * clamp01((base_b * shade)))));
                    } else {
                        const tsky: f64 = (0.5 * (dy + 1.0));
                        r = @as(i64, @intFromFloat((255.0 * (0.65 + (0.2 * tsky)))));
                        g = @as(i64, @intFromFloat((255.0 * (0.75 + (0.18 * tsky)))));
                        b = @as(i64, @intFromFloat((255.0 * (0.9 + (0.08 * tsky)))));
                    }
                    ar += r;
                    ag += g;
                    ab += b;
                }
            }
            const samples: i64 = (aa * aa);
            pytra.list_append(pixels, u8, @intCast(@divFloor(ar, samples)));
            pytra.list_append(pixels, u8, @intCast(@divFloor(ag, samples)));
            pytra.list_append(pixels, u8, @intCast(@divFloor(ab, samples)));
        }
    }
    return pixels;
}

fn run_raytrace() void {
    const width: i64 = 1600;
    const height: i64 = 900;
    const aa: i64 = 2;
    const out_path: []const u8 = "sample/out/02_raytrace_spheres.png";
    
    const start: f64 = pytra.perf_counter();
    const pixels: pytra.Obj = render(width, height, aa);
    png.write_rgb_png(out_path, width, height, pixels);
    const elapsed: f64 = (pytra.perf_counter() - start);
    
    pytra.print2("output:", out_path);
    pytra.print("size:");
    pytra.print2("elapsed_sec:", elapsed);
}

pub fn main() void {
    run_raytrace();
}
